ecosystem/typescript/sdk_v2/src/crypto/asymmetric_crypto.ts

```diff
@@ -1,14 +1,14 @@
 // Copyright © Aptos Foundation
 // SPDX-License-Identifier: Apache-2.0
 
-import { Deserializable, Deserializer, Serializable, Serializer } from "../bcs";
+import { Serializable, Serializer } from "../bcs";
 import { HexInput } from "../types";
 
 /**
  * An abstract representation of a public key.  All Asymmetric key pairs will use this to
  * verify signatures and for authentication keys.
  */
-export abstract class PublicKey implements Serializable, Deserializable<PublicKey> {
+export abstract class PublicKey extends Serializable {
```

Comment:
I see! In this case, the implementation of the method `deserialize` won't be enforced for child classes?

I wonder if theres a way to just implement the deserialize method directly in abstract/base class. Then all child classes will auto inherit it

Reply:
Yeah I looked into it and there's apparently a lot of hullabaloo about static methods in abstract classes, so the support there, particularly with typescript, is lacking.

I tried a few different methods to achieve a `static deserialize` method interface inherited from the abstract class but it came down to either being really janky or just not possible.

WRT what you're saying though, the problem with implementing the deserialize method in the base class (say, as an abstract factory) is that it must know ahead of time how to create the child classes. I don't know if this would be a problem, it depends on how we plan on implementing these abstract classes later.

For example here is the `TransactionAuthenticator` abstract factory pattern:

```typescript
export abstract class TransactionAuthenticator {
  abstract serialize(serializer: Serializer): void;

  static deserialize(deserializer: Deserializer): TransactionAuthenticator {
    const index = deserializer.deserializeUleb128AsU32();
    switch (index) {
      case 0:
        return TransactionAuthenticatorEd25519.load(deserializer);
      case 1:
        return TransactionAuthenticatorMultiEd25519.load(deserializer);
      case 2:
        return TransactionAuthenticatorMultiAgent.load(deserializer);
      case 3:
        return TransactionAuthenticatorFeePayer.load(deserializer);
      default:
        throw new Error(`Unknown variant index for TransactionAuthenticator: ${index}`);
    }
  }
}
```

But with this, the child classes actually don't even implement deserialize, they specifically have a `load` function they use.

The other part to this is that these are specifically `enum`s in Rust, so that's why they're serialized/deserialized like this.

With the `PublicKey` stuff, it's not an enum in rust, it's merely an abstract class we're using for the SDK, so we wouldn't be serializing an enum value (`0`, `1`, etc) at the beginning of the byte buffer to differentiate between child class implementations.